File: src/main/java/core/procedures/strings/StringFill.java
package core.procedures.strings;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.MutableString;

public final class StringFill extends AFn {

  public StringFill() {
    super(new FnArgsBuilder().min(2).max(2)
                             .mandatory(new Class[]{MutableString.class, Character.class}).build());
  }

  @Override
  public String getName() {
    return "string-fill!";
  }

  @Override
  public boolean isPure() {
    return false;
  }

  @Override
  public MutableString apply2(Object arg1, Object arg2) {
    MutableString s = (MutableString)arg1;
    int oldLength = s.length();
    s.clear();
    for (int i = 0; i < oldLength; i++) {
      s.append(arg2);
    }
    return s;
  }
}

File: src/main/java/core/procedures/strings/StringFill.kt
package core.procedures.strings

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.MutableString

class StringFill : AFn<Any?, MutableString>(name = "string-fill!", arity = Exactly(2),
                                            mandatoryArgsTypes = arrayOf(MutableString::class.java,
                                                                         Char::class.javaObjectType)) {

    override operator fun invoke(arg1: Any?, arg2: Any?): MutableString {
        val s = arg1 as MutableString?
        val oldLength = s!!.length
        return s.apply {
            clear()
            for (i in 0 until oldLength) { append(arg2) }
        }
    }
}
